synapse/handlers/room.py

```diff
@@ -215,6 +218,9 @@ async def _upgrade_room(
 
         old_room_state = await tombstone_context.get_current_state_ids()
 
+        # We know the tombstone event isn't an outlier so it has current state.
+        assert old_room_state is not None
```

Comment:
I wonder if we should warn people that running Synapse with python's optimise (`-O`) option will strip these?

Reply:
Warn who? If you're running it with `-O` then you already know that asserts should get removed. Besides, this should never be hit.